Add channel attr

KamranHussain05 opened the pull request:
Added a few functions to derive a channel-level raster (instead of neurons x time) and a frame-level raster (instead of ms). This can serve as a nice training mask that is derived from existing data, so it backwards compatible with previous spike data verisons.

KamranHussain05 committed:
add neuron to channel raster and frame level raster (mask of spikes over frames for each channel

KamranHussain05 committed:
add automatic channel mapping resolution

KamranHussain05 committed:
update channel raster

KamranHussain05 committed:
update raster

atspaeth commented on spikedata/spikedata.py:
```diff
@@ -524,6 +524,424 @@ def raster(self, bin_size=20.0):
+            current_channels = channel_raster.shape[0]
+            if current_channels < expected_num_channels:
+                # Pad with zeros
+                if sparse.issparse(channel_raster):
```
What's the point of the sparse path if it involves creating a dense temp array? At a glance I don't see anything in the dense code that's more memory-intensive than that

atspaeth commented on spikedata/spikedata.py:
```diff
@@ -524,6 +524,424 @@ def raster(self, bin_size=20.0):
+        neuron_raster = self.sparse_raster(bin_size)
+
+        # Aggregate by channel
+        if sparse.issparse(neuron_raster):
```
Won't it always be sparse since it comes straight from `sparse_raster()`?

atspaeth commented on spikedata/spikedata.py:
```diff
@@ -1234,23 +1652,22 @@ def butter_filter(
+            raise ValueError("lowcut must be smaller than highcut")
+        filter_type = "bandpass"
+        band = [lowcut, highcut]
+        Wn = [e / fs * 2 for e in band]
```
Why did you need to translate everything back to Python 3.9 syntax?

atspaeth reviewed (approved):
This looks really useful and if you want you can merge already, but I do recommend getting rid of the sparse option, or at least simplifying the `issparse` check logic.

KamranHussain05 committed:
address review comments, remove sparse data check

KamranHussain05 commented:
Sorry, accidentally closed lol

KamranHussain05 commented:
@atspaeth Most recent push addressed your comments and simplifies the sparsity checking code. Lmk wyt! Also CIs are failing suddenly due to Numpy version error, can constrain this to `numpy<2.0` or update CIs--likely a diff PR.

atspaeth commented:
I think your fixes look good, and the unit tests pass locally.

Don't add that version guard though. The CI issue is something weirder. It's running Python 3.11 instead of the intended version, and it claims to be installing numpy 2.3.5 but then fails to parse 2.4.0rc1 (which is a prerelease ofc and shouldn't be installable without `--pre` in pip)

KamranHussain05 commented:
Good to merge?

atspaeth commented:
I'm going to try to fix that CI issue in a separate branch
